### DISCUSSION TITLE
Reading model issues and an idea

### DISCUSSION THREAD

**Discussion by bakerwy** [fumitoh/modelx#165] [Ideas]
In #153 , I mentioned reading model consumes quite long time and large memory (Almost 7GB for memory and 600 sec for opening model)

So, I tested memory and time of reading data.

The most memory consumed modules were as follows:

C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\tokenize.py:537
C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\ast.py
C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\line_numbers.py

The most time consumed module was as follows:

C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site- packages\asttokens\mark_tokens
C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\asttokens.py
C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\mark_tokens.py

I didn't use external data as pickle. 

It seems if there are a lot of complex functions, lots of resources are consumed to transfer the code to byte stream and tokenize them (not sure though).

Thinking of practices, for the maintenance purpose, we open specific spaces not all the spaces in the model at the same time.
so, I think it would be great if we can open specific spaces and operate them and save.

**Comment by fumitoh** [comment]
I think it's difficult to only read some part of a model. Maybe the speed of reading a model can be improved.
I want to know whether asttokens is slow.

The code below tests how much time asttokens spends to parse all the souce files in a model.
Can you run the code on your model and see how much it takes?

```python
import time
from pathlib import Path
import asttokens

def read_init_files(directory):
    init_files_content = {}  # Dictionary to store file paths and their contents
    
    # Convert the input directory to a Path object
    directory = Path(directory)
    
    # Walk through the directory structure
    for init_file_path in directory.rglob('__init__.py'):
        try:
            # Read the contents of the __init__.py file
            init_files_content[init_file_path] = txt = init_file_path.read_text()
            atok = asttokens.ASTTokens(txt, parse=True)
            
            for i, stmt in enumerate(atok.tree.body):
                atok.get_text(stmt)   # dummpy operation
            
        except Exception as e:
            print(f"Error reading file {init_file_path}: {e}")
    
    return init_files_content

if __name__ == "__main__":
    directory = input("Enter the model directory path to scan: ")

    start_time = time.time()
    
    init_files_content = read_init_files(directory)

    end_time = time.time()
    
    print(f"Script run time: {end_time - start_time:.4f} seconds")
    
    # for file_path, content in init_files_content.items():
    #     print(f"{file_path}: {len(content)} characters")
```

**Reply by bakerwy** [reply]
I used the code you suggested and it took 98.5416 seconds for the heavy model.

**Reply by bakerwy** [reply]
And here is the top 30 of cProfile **(total 1146 sec)**

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 21088977  106.369    0.000  242.849    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\mark_tokens.py:127(_expand_to_matching_pairs)
314147161   76.378    0.000  111.744    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\asttokens.py:242(token_range)
 17966721   55.685    0.000  437.110    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\mark_tokens.py:76(_visit_after_children)
   112405   53.457    0.000  704.841    0.006 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\util.py:233(visit_tree)
101691588   52.317    0.000  106.799    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\ast.py:269(iter_child_nodes)
   337353   45.817    0.000   45.817    0.000 {built-in method builtins.compile}
 79188538   44.437    0.000  160.692    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\util.py:148(iter_children_ast)
 31658566   43.985    0.000   43.985    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\util.py:381(<genexpr>)
 80982339   39.656    0.000   71.314    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\line_numbers.py:53(line_to_offset)
307791735   36.552    0.000   36.552    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\util.py:100(is_non_coding_token)
 17702486   34.579    0.000   57.514    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\line_numbers.py:37(from_utf8_col)
 31658566   34.041    0.000  237.114    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\asttokens.py:144(_translate_tokens)
136855901   28.759    0.000   39.686    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\ast.py:257(iter_fields)
 31658566   26.914    0.000   49.389    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\tokenize.py:530(_generate_tokens_from_c_tokenizer)
 31658566   25.273    0.000  136.351    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\asttokens\util.py:354(patched_generate_tokens)
   262341   23.275    0.000   57.644    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\ast.py:34(parse)
214098942   20.909    0.000   21.960    0.000 {built-in method builtins.isinstance}
153398658/153247102   16.186    0.000   16.734    0.000 {built-in method builtins.len}
 98298245   15.512    0.000   15.513    0.000 {built-in method builtins.max}
 98298679   14.871    0.000   14.871    0.000 {built-in method builtins.min}
116742405   14.234    0.000   14.239    0.000 {built-in method builtins.getattr}
 31546161   13.203    0.000   22.475    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\collections\__init__.py:447(_make)
 87080334   12.679    0.000   12.679    0.000 {method 'pop' of 'list' objects}
64146035/64145962   11.856    0.000   11.857    0.000 {built-in method __new__ of type object at 0x00007FF8111398B0}
    75420   11.654    0.000   22.684    0.000 {built-in method builtins.all}
 86321430   11.356    0.000   11.356    0.000 {method 'append' of 'list' objects}
193715576   11.030    0.000   11.030    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\modelx\core\base.py:624(<genexpr>)
26957874/26919732   10.877    0.000   60.710    0.000 {built-in method builtins.any}
 97663328   10.625    0.000   10.625    0.000 {method 'encode' of 'str' objects}
    75958   10.078    0.000   25.321    0.000 C:\Users\LG\AppData\Local\Programs\Python\Python312\Lib\site-packages\modelx\core\base.py:847(_update_interfaces)
    
```

**Comment by fumitoh** [comment]
I think by optimizing modelx's code, your model could be loaded in 120 - 150 seconds.
To make it load faster than that, maybe free-threading can be utilized in future.

If you only want to change the fomula of a cells, you can just edit it in the source `__init__.py` file.
Another thing you can try is to utilize inheritance to eliminate duplicated formulas.

**Reply by bakerwy** [reply]
I understand the idea free thread and inheritance. but I don't understand modelx's code optimization and editing __init__.py

Can you give me some examples or idea? thanks in advance

**Reply by fumitoh** [reply]
I meant by editing the source `__init__.py`, Cells formulas ares saved in `__init__.py` file under the directory whose name is the parent space's name in the model directory, so you can edit cells formulas without reading the model.

I meant by modelx's code optimization to improve modelx itself by rewriting code related to model reading and creation, so this is for myself to work on.

**Reply by bakerwy** [reply]
Ahh Thanks fumito!

**Reply by fumitoh** [reply]
I made some updates to modelx to make `read_model` faster. Can you test the development version of modelx to see how long it takes to load your model? You can install the dev version of modelx by `pip -e` as explained here:https://stackoverflow.com/a/35064498/2400267

**Reply by bakerwy** [reply]
I tested it and it took 165 seconds! (Originally it took 600 seconds)
Is it your target. And also memory usage reduced (6.8GB. Originally 7.8GB).

How did you make efficient code?

**Reply by fumitoh** [reply]
asttokens was taking time, so I made changes to minimize the use of asttokens in modelx as much as possible.
I'm making further changes  to almost eliminate the use of asttokens, which will be reflected to the repository in a week or so.

**Reply by fumitoh** [reply]
I made furthur improvements in modelx for reading large models. Can you test the following version with your model and measure the loading time?

https://github.com/fumitoh/modelx/tree/feat-fast-read

**Reply by bakerwy** [reply]
Wow.. it took only 73 seconds and it consumes only 500 mega bytes while reading. 
What is the changes you made?

**Reply by bakerwy** [reply]
Can you make export faster too? Applying your method for that?

**Reply by fumitoh** [reply]
How long does it take to export your model?

**Reply by bakerwy** [reply]
I just tested it and it took 23 minute for exporting

**Reply by fumitoh** [reply]
export uses another library, libcst. export is more complex than read, so it's difficult to make it faster at the moment. Free-threading may be utilized in future.
